content/blog/copilot-lessons/index.md

```diff
@@ -0,0 +1,129 @@
+
+## Testing the Untestable: Validating LLM Outputs
+
+When testing traditional code, we expect consistent, predictable outputs. With LLMs, even correct answers can vary. Here's how we tackle this challenge.
+
+Our first approach was simple: keyword checks. For example, when testing the update analysis feature, we checked if the LLM’s response included the word “security” and described the error. This worked for straightforward cases but broke down quickly. Take a question like, “How many Lambdas am I running?” The LLM might give the right numbers but skip the word “running,” failing the test even though the answer was correct.
+
+These early failures revealed the limitations of keyword-based validation and underscored the need for a more nuanced approach. Inspired by platforms like LangSmith and Promptfoo, we began leveraging LLMs themselves as evaluators. For deterministic tasks, simple keyword checks suffice, but for more complex scenarios—like assessing whether a response answers a specific question—we rely on an "LLM Judge." This approach balances efficiency and flexibility, reserving LLM evaluation for cases where it truly adds value. Our test suite now integrates both methods, running against every code change to validate response content, accuracy, and format.
+
+![Example PromptFoo Eval](promptfoo.png)
+
+This LLM-as-judge approach significantly improved in our ability to catch subtle errors that keyword checking had missed.
+
+These early failures showed the limits of keyword-based validation and the need for a better approach. Inspired by tools like LangSmith and Promptfoo, we started using LLMs for validation. Simple checks work well for straightforward cases, but for harder questions—like whether a response answers what the user asked—we use an LLM Judge. This way, we only rely on LLMs when they’re really needed. Our test suite now uses both methods to check every code change, making sure responses are accurate and high quality.
+
+Our eval suite keeps getting more robust, whic means that when new AI models drop, we can quickly catch any weirdness before it hits production. The generative AI space moves crazy fast and our code changes a lot, but our evals are our safety net - catching hallucinations, maintaining quality, and making sure we don't ship anything that'll annoy our users.
+
+So while hallucinations are now much more rare, there is that one I can't stop thinking about - that `--force` flag incident. Not because it was a bug, but because it taught us something fascinating about these AI errors.
+
+## Errors Point the Way
+
+The `--force` hallucination wasn't totally wrong - it was revealing what users intuitively expect from our CLI. Now I'm pushing for getting that flag added, and all because our LLM accidentally showed us what was missing. Force deletion is a common pattern across developer tools, and the LLM, trained on vast amounts of documentation and code, simply reflects these established conventions.
```

Comment:
> Now I'm pushing for

Let's not say that. If you follow Fraser's response, we actually considered that -- a bug was opened earlier.
How about something like this:

"We've considered adding the flag before but weren't sure if that's the semantics the user would expect - now we get more customer signal to help us make the right call".